File: test_bayesian_multipara.jl
using LinearAlgebra
using Trapz
using Interpolations
using PyCall

include("quanestimation/JuliaSrc/Common/common.jl")
include("quanestimation/JuliaSrc/Dynamics/dynamics.jl")
include("quanestimation/JuliaSrc/AsymptoticBound/CramerRao.jl")
include("quanestimation/JuliaSrc/BayesianBound/BayesianCramerRao.jl")
include("quanestimation/JuliaSrc/BayesianBound/ZivZakai.jl")

#initial state
rho0 = 0.5*[1.0 1.0+0.0im; 1.0 1.0]
#Hamiltonian
# B = pi/2.0
sx = [0.0 1.0; 1.0 0.0im]
sy = [0.0 -1.0im; 1.0im 0.0]
sz = [1.0 0.0im; 0.0 -1.0]

function H0_res(x,B)
    return 0.5*B*(sx*cos(x)+sz*sin(x))
end
function dH_res(x,B)
    return [0.5*B*(-sx*sin(x)+sz*cos(x)), 0.5*(sx*cos(x)+sz*sin(x))]
end

#measurement
Measurement = sic_povm([0.37637620719571985+2.7760878126176896im,1.1538819157681195+0.87835540934078105im])

tspan = range(0.0, stop=1.0, length=1000)

dim = 2
para_num = 2

decay_opt = [zeros(ComplexF64,para_num,para_num)] 
γ = [0.0]

#### flat distribution ####
xspan = [range(-pi/2.0, stop=pi/2.0, length=100), range(pi/2-0.01, stop=pi/2+0.01, length=10)].|>Vector
x1, x2 = [x[1] for x in xspan],  [x[end] for x in xspan]

σ_x = 0.5
σ_B = 1.0
# ρ = 0.5
μ = [0., 0.]

BQCRB1 =  Matrix{Float64}[]
BQCRB2 =  Matrix{Float64}[]
QVTB1 =  Matrix{Float64}[]
QVTB2 =  Matrix{Float64}[]
BCRB1 =  Matrix{Float64}[]
BCRB2 =  Matrix{Float64}[]
VTB1 =  Matrix{Float64}[]
VTB2 =  Matrix{Float64}[]

for ρ in range(0.1,0.9,length=10)
    BivarNormal(x,B) = 1/(2pi*σ_x*σ_B*sqrt(1-ρ^2))*exp(-((x-μ[1])^2/σ_x^2-2ρ*(x-μ[1])*(B-μ[2])/σ_x/σ_B + (B-μ[2])^2/σ_B^2)/2/(1-ρ^2))
    function dBivarNormal(x,B)
        BN = BivarNormal(x,B)
        [BN*(2*(x-μ[1])/σ_x^2-2*ρ*(B-μ[2])/σ_B/σ_x), BN*(2*(B-μ[2])/σ_B^2-2*ρ*(x-μ[1])/σ_x/σ_B)]./(-2*(1- ρ^2))
    end
    p = [BivarNormal(x,B) for (x,B) in Iterators.product(xspan...)]
    trapzm(x, integrands, slice_dim) =  [trapz(tuple(x...), I) for I in [reshape(hcat(integrands...)[i,:], length.(x)...) for i in 1:slice_dim]]
    C = trapzm(xspan, p, 1)
    p = p./C
    dp = Matrix{Vector{Float64}}(undef, length.(xspan)...)
    rho_all = Matrix{Matrix{ComplexF64}}(undef, length.(xspan)...)
    drho_all = Matrix{Vector{Matrix{ComplexF64}}}(undef, length.(xspan)...)
    for i = 1:length(xspan[1])
        for j = 1:length(xspan[2])  
            H0 = H0_res(xspan[1][i], xspan[2][j])
            dH = dH_res(xspan[1][i], xspan[2][j])
            rho, drho  = expm(H0, dH, rho0, decay_opt, γ, [zeros(ComplexF64,2,2)], [zeros(length(tspan)-1)],tspan)
            rho_all[i,j] = rho[end]
            drho_all[i,j] = drho[end]
            dp[i,j] = dBivarNormal(xspan[1][i],xspan[2][j])./C
        end
    end
    append!(BCRB1, [BCRB(xspan, p, rho_all, drho_all, M=Measurement, eps=1e-8)])
    append!(BCRB2, [BCRB(xspan, p, rho_all, drho_all, M=Measurement, btype=2, eps=1e-8)])
    append!(BQCRB1, [BQCRB(xspan, p, rho_all, drho_all, eps=1e-8)])
    append!(BQCRB2, [BQCRB(xspan, p, rho_all, drho_all, btype=2, eps=1e-8)])
    append!(VTB1, [VTB(xspan, p, dp, rho_all, drho_all, M=Measurement, eps=1e-8)])
    append!(VTB2, [VTB(xspan, p, dp, rho_all, drho_all, btype=2, M=Measurement, eps=1e-8)])
    append!(QVTB1, [QVTB(xspan, p, dp, rho_all, drho_all, eps=1e-8)])
    append!(QVTB2, [QVTB(xspan, p, dp, rho_all, drho_all, btype=2, eps=1e-8)])
end

np = pyimport("numpy")
np.save("BCRB1.npy",BCRB1)
np.save("BCRB2.npy",BCRB2)
np.save("BQCRB1.npy",BQCRB1)
np.save("BQCRB2.npy",BQCRB2)
np.save("VTB1.npy",VTB1)
np.save("VTB2.npy",VTB2)
np.save("QVTB1.npy",QVTB1)
np.save("QVTB2.npy",QVTB2)

File: test_bayesian_multipara.py
import numpy as np
from quanestimation import *
from scipy import integrate

def p_gauss(x, y, mu_x, mu_y, sigmax, sigmay, r):
    term1 = ((x-mu_x)/sigmax)**2-2*r*(((x-mu_x)/sigmax))*(((y-mu_y)/sigmay))+(((y-mu_y)/sigmay))**2
    term2 = np.exp(-term1/2.0/(1-r**2))
    term3 = 2*np.pi*sigmax*sigmay*np.sqrt(1-r**2)
    return term2/term3

def dxp_gauss(x, y, mu_x, mu_y, sigmax, sigmay, r):
    term = -(2*((x-mu_x)/sigmax**2)-2*r*((y-mu_y)/sigmay)/sigmax)/2.0/(1-r**2)
    return term*p_gauss(x, y, mu_x, mu_y, sigmax, sigmay, r)

def dyp_gauss(x, y, mu_x, mu_y, sigmax, sigmay, r):
    term = -(2*((y-mu_y)/sigmay**2)-2*r*((x-mu_x)/sigmax)/sigmay)/2.0/(1-r**2)
    return term*p_gauss(x, y, mu_x, mu_y, sigmax, sigmay, r)

#initial state
rho0 = 0.5*np.array([[1., 1.],[1., 1.]])
#Hamiltonian
B = np.pi/2.0
sx = np.array([[0., 1.],[1., 0.0j]])
sy = np.array([[0., -1.j],[1.j, 0.]]) 
sz = np.array([[1., 0.0j],[0., -1.]])
H0_res = lambda x, B: 0.5*B*(sx*np.cos(x)+sz*np.sin(x))
dH1_res = lambda x, B: 0.5*B*(-sx*np.sin(x)+sz*np.cos(x))
dH2_res = lambda x, B: 0.5*(sx*np.cos(x)+sz*np.sin(x))

#measurement
M1 = 0.5*np.array([[1., 1.],[1., 1.]])
M2 = 0.5*np.array([[1.,-1.],[-1., 1.]])
M = [M1, M2]

T = 1.0
tnum = int(1000*T)
tspan = np.linspace(0., T, tnum)

dim = len(rho0)
para_num = 2

### gaussian distribution
xspan = np.linspace(-0.5*np.pi, 0.5*np.pi, 100)
yspan = np.linspace(0.5*np.pi-0.01, 0.5*np.pi+0.01, 10)
sigmax, sigmay = 0.5, 1.0
mu_x, mu_y = 0.0, 0.0
x1, x2 = xspan[0], xspan[-1]
y1, y2 = yspan[0], yspan[-1]
rho_span = np.linspace(0.1,0.9,10)
f_BCRB1, f_BCRB2, f_VTB1, f_VTB2 = [], [], [], []
f_BQCRB1, f_BQCRB2, f_QVTB1, f_QVTB2 = [], [], [], []
f_QZZB = []
for r in rho_span:
    p_tp = [[p_gauss(xspan[j], yspan[i], mu_x, mu_y, sigmax, sigmay, r) for i in range(len(yspan))] for j in range(len(xspan))]
    dp_tp = [[[0.0, 0.0] for j in range(len(yspan))] for k in range(len(xspan))]

    for ii in range(len(xspan)):
        for jj in range(len(yspan)):
            dp_tp[ii][jj][0] = dxp_gauss(xspan[ii], yspan[jj], mu_x, mu_y, sigmax, sigmay, r)
            dp_tp[ii][jj][1] = dyp_gauss(xspan[ii], yspan[jj], mu_x, mu_y, sigmax, sigmay, r)

    S = integrate.simps(integrate.simps(p_tp, yspan), xspan)
    p = p_tp/S
    dp = dp_tp/S

    rho_all = [[np.zeros((dim, dim), dtype=np.complex128) for i in range(len(yspan))] for j in range(len(xspan))]
    drho_all = [[[np.zeros((dim, dim), dtype=np.complex128) for i in range(para_num)] for j in range(len(yspan))] for k in range(len(xspan))]
    for i in range(len(xspan)):
        for j in range(len(yspan)):
            H0 = H0_res(xspan[i], yspan[j])
            dH1 = dH1_res(xspan[i], yspan[j])
            dH2 = dH2_res(xspan[i], yspan[j])
            dynamics = Lindblad(tspan, rho0, H0, [dH1,dH2])
            rho_tp, drho_tp = dynamics.expm()
            rho, drho = rho_tp[-1], drho_tp[-1]
            rho_all[i][j] = rho
            for k in range(para_num):
                drho_all[i][j][k] = drho[k]
    f0_1 = BCRB([xspan, yspan], p, rho_all, drho_all, M=[], btype=1, eps=1e-8)
    f0_2 = BCRB([xspan,yspan], p, rho_all, drho_all, M=[], btype=2, eps=1e-8)
    f1_1 = BQCRB([xspan,yspan], p, rho_all, drho_all, btype=1, eps=1e-8)
    f1_2 = BQCRB([xspan,yspan], p, rho_all, drho_all, btype=2, eps=1e-8)
    f_BCRB1.append(f0_1)
    f_BCRB2.append(f0_2)
    f_BQCRB1.append(f1_1)
    f_BQCRB2.append(f1_2)

    f2_1 = VTB([xspan,yspan], p, dp, rho_all, drho_all, M=[], btype=1, eps=1e-8)
    f2_2 = VTB([xspan,yspan], p, dp, rho_all, drho_all, M=[], btype=2, eps=1e-8)
    f3_1 = QVTB([xspan,yspan], p, dp, rho_all, drho_all, btype=1, eps=1e-8)
    f3_2 = QVTB([xspan,yspan], p, dp, rho_all, drho_all, btype=2, eps=1e-8)
    f_VTB1.append(f2_1)
    f_VTB2.append(f2_2)
    f_QVTB1.append(f3_1)
    f_QVTB2.append(f3_2)

np.save("f_BCRB1", f_BCRB1)
np.save("f_BCRB2", f_BCRB2)
np.save("f_BQCRB1", f_BQCRB1)
np.save("f_BQCRB2", f_BQCRB2)
np.save("f_VTB1", f_VTB1)
np.save("f_VTB2", f_VTB2)
np.save("f_QVTB1", f_QVTB1)
np.save("f_QVTB2", f_QVTB2)    

#### flat distribution ####
# xspan = np.linspace(-0.5*np.pi, 0.5*np.pi, 100)
# yspan = np.linspace(0.5*np.pi-0.1, 0.5*np.pi+0.1, 10)
# x1, x2 = xspan[0], xspan[-1]
# y1, y2 = yspan[0], yspan[-1]
# p = (1.0/(x2-x1))*(1.0/(y2-y1))*np.ones((len(xspan), len(yspan)))
# dp = [[[0.0,0.0] for i in range(len(yspan))] for j in range(len(xspan))]
# rho_all = [[np.zeros((dim, dim), dtype=np.complex128) for i in range(len(yspan))] for j in range(len(xspan))]
# drho_all = [[[np.zeros((dim, dim), dtype=np.complex128) for i in range(para_num)] for j in range(len(yspan))] for k in range(len(xspan))]
# for i in range(len(xspan)):
#     for j in range(len(yspan)):
#         H0 = H0_res(xspan[i], yspan[j])
#         dH1 = dH1_res(xspan[i], yspan[j])
#         dH2 = dH2_res(xspan[i], yspan[j])
#         dynamics = Lindblad(tspan, rho0, H0, [dH1, dH2])
#         rho_tp, drho_tp = dynamics.expm()
#         rho, drho = rho_tp[-1], drho_tp[-1]
#         rho_all[i][j] = rho
#         for k in range(para_num):
#             drho_all[i][j][k] = drho[k]

# # fc = BCFIM([xspan, yspan], p, rho_all, drho_all, M=[])
# # fq = BQFIM([xspan, yspan], p, rho_all, drho_all)
# f0_1 = BCRB([xspan, yspan], p, rho_all, drho_all, M=[], btype=1, eps=1e-8)
# f0_2 = BCRB([xspan, yspan], p, rho_all, drho_all, M=[], btype=2, eps=1e-8)
# f1_1 = BQCRB([xspan, yspan], p, rho_all, drho_all, btype=1, eps=1e-8)
# f1_2 = BQCRB([xspan, yspan], p, rho_all, drho_all, btype=2, eps=1e-8)
# print(f0_1, f0_2, f1_1, f1_2)
# f2_1 = VTB([xspan, yspan], p, dp, rho_all, drho_all, M=[], btype=1, eps=1e-8)
# f2_2 = VTB([xspan, yspan], p, dp, rho_all, drho_all, M=[], btype=2, eps=1e-8)
# f3_1 = QVTB([xspan, yspan], p, dp, rho_all, drho_all, btype=1, eps=1e-8)
# f3_2 = QVTB([xspan, yspan], p, dp, rho_all, drho_all, btype=2, eps=1e-8)
# print(f2_1, f2_2, f3_1, f3_2)
